### ISSUE TITLE
Network not learning

### ISSUE THREAD

**Issue by S-Siva** [dnouri/nolearn#251]
I am trying to use CNN for classification, and I seem to have a bit of a problem.
My network right now:
    net = NeuralNet(
    layers=[
        ('input', layers.InputLayer),
        ('conv1', layers.Conv1DLayer),
        ('pool1', layers.MaxPool1DLayer),
        ('conv2', layers.Conv1DLayer),
        ('pool2', layers.MaxPool1DLayer),
        ('conv3', layers.Conv1DLayer),
        ('pool3', layers.MaxPool1DLayer),
        ('hidden4', layers.DenseLayer),
        ('hidden5', layers.DenseLayer),
        ('output', layers.DenseLayer),
        ],
    input_shape=(None,1, features), 
    conv1_num_filters=10, conv1_filter_size=(15),
    pool1_pool_size=(5),
    conv2_num_filters=5, conv2_filter_size=(25),
    pool2_pool_size=(2),
    conv3_num_filters=2, conv3_filter_size=(2), pool3_pool_size=(1),
    hidden4_num_units=1000,
    hidden5_num_units=500,
    output_num_units=8,
    output_nonlinearity=nonlinearities.softmax,

    #update=nesterov_momentum,
    update_learning_rate=0.03,
    #update_momentum=0.9,
    regression=True,
    max_epochs=100,
    verbose=1,
    )
The number of inputs is ~20000 and outputs is 8
When using the net to predict data from the test set I get:
Actual [ 0.12134051  0.1198738   0.11978325  0.11679561  0.12244498  0.13342009
  0.13341455  0.13292722], Expected [1,0,0,0,0,0,0,0]
Actual [ 0.12134027  0.11987287  0.11978108  0.11679234  0.12244509  0.13342562
  0.13341345  0.1329293 ], , Expected [0,0,0,0,0,0,0,1]
I have it tried with and without momentum, different number of layers, different learning rates etc.
I have 3 folds and when looking at the confusion matrix, I get 3 columns completely filled and the other 5 always 0 i.e.
              Predicted Class
[[  20.    0.    0.    0.    0.   14.   24.    0.]
 [  35.    0.    0.    0.    0.   28.   35.    0.]        
 [  55.    0.    0.    0.    0.   53.   59.    0.]
 [  73.    0.    0.    0.    0.   81.   94.    0.]        Actual Class
 [  86.    0.    0.    0.    0.   80.   92.    0.]
 [ 100.    0.    0.    0.    0.   90.  108.    0.]
 [ 154.    0.    0.    0.    0.  152.  155.    0.]
 [ 162.    0.    0.    0.    0.  133.  171.    0.]]
It looks like the network is predicting only one class for each fold.

The problem is that when I tried the same network architecture yesterday on the same machine with the same train and test set, it predicted all the classes despite not being very accurate.

**Comment by BenjaminBossan** [comment]
Could you try whether setting `regression=False` solves the problem?

**Comment by S-Siva** [comment]
I did try that too, it just predicts one class then too.

**Comment by BenjaminBossan** [comment]
Regardless, you should set the parameter to False, since you have a classification problem.

It is difficult to tell what goes wrong here. I would start with a very simple architecture and check whether the net learns anything at all. What does the loss function say (set `verbose=1`)?

If the loss does not decrease, try random data that has the same shape as your real data, usually a neural net should be able to overfit these data. Be patient when training, the net may take several epochs before the loss starts decreasing.

Lastly, try to check whether you can find any difference between the setting that worked and your current one. Often, there are subtle differences that matter.
